### PR TITLE
Code Flag Options

### LINKED ISSUE #1801: --code doesn't work with --check or --diff
**Describe the bug**

- `--code` is supposed to take the code as an input string
- `--check` is supposed to change the exit code based on whether edits were made while formatting
- `--diff` is supposed to only output a diff, instead of the entire formatted source

What I found was the the `--check` and `--diff` flags did not work when passing input using `--code`; it behaved as if the extra flags hadn't been passed at all.

**To Reproduce** Steps to reproduce the behavior:

1. Run this command (notice the double space after `=`):

```sh
black --check --code 'a =  5'
```

2. The command exits with a status `0`, despite making edits (removing the double space).

**Expected behavior** The command should have exited with a status of `1`, since edits were made.

**Environment (please complete the following information):**

- Version: `20.8b1`
- OS and Python version: Arch Linux/Python 3.8.6

### LINKED ISSUE #2104: Safety checks are disabled when passing in code using `-c` / `--code`
I've been writing some docs during my documentation reorganization effort, and I've found myself writing this warning:

> --check, --diff, and --safe / --fast have no effect when using -c / --code. Safety checks normally turned on by default that verify Black's output are disabled as well.

~Is this by design? IMO it seems like an oversight / bug that should fixed.~

edit: yep this is in fact a bug, [thanks Jelle](https://github.com/psf/black/issues/2104#issuecomment-818302531)!

cc @JelleZijlstra @cooperlees

### PR DESCRIPTION
**Pull request by HassanAbouelela**
Closes #2104, closes #1801.

Fixes a bug in the `--code` flag which would ignore options such as `--diff`, `--check`, and `--fast`.

This PR implements the fix described in the issue. It also adds tests for this fix.

### CODE REVIEW

**Comment by HassanAbouelela** [comment]
I've fixed the linting issues now, could someone rerun the workflows.

**Review by cooperlees — APPROVED** [review]
Wow, great tests here! Thanks for your contribution.

File: src/black/__init__.py
Comment (ichard26) [inline src/black/__init__.py:406]: I don't know whether `echo` is better than `print` here (and we use both inconsistently in the codebase) but `echo` seems to be safer?

**Review by ichard26 — COMMENTED** [review]
Thanks for working on this fix! I got some feedback but nothing truly blocking (except for possibly a Windows bug).

Great tests btw!

File: src/black/__init__.py
Comment (HassanAbouelela) [reply src/black/__init__.py:32]: You'd be right for the most part, but `diff` is a parameter of the main function, and overrides the global diff function. I figured doing it this way would avoid changing the interface of the function.

File: src/black/__init__.py
Comment (HassanAbouelela) [reply src/black/__init__.py:406]: That may be the case, but if we're going to be using a textiowrapper on stdout, we could just use the write method, as is done in the linked example.

Comment (ichard26) [reply src/black/__init__.py:406]: *facepalm* should've caught that, my bad

File: src/black/__init__.py
Comment (HassanAbouelela) [reply src/black/__init__.py]: I completely missed that, thanks.

Resolved in 6221b880.

File: src/black/__init__.py
Comment (HassanAbouelela) [reply src/black/__init__.py]: No longer relevant. 6221b88.

File: tests/test_black.py
Comment (HassanAbouelela) [reply tests/test_black.py]: Thanks, I've updated the tests. 081a8e32.

File: tests/test_black.py
Comment (ichard26) [inline tests/test_black.py:2084]: Nice clean up!

**Review by ichard26 — CHANGES_REQUESTED** [review]
This time I did a hands-on review / test and this still isn't quite right. The error message when formatting fails isn't right. For normal cases, this is handled by the `Report` class. Check out the difference between these cases:

`--code`:

Screenshot from 2021-05-27 18-51-43

STDIN:

Screenshot from 2021-05-27 18-50-41

It honestly might be better to generalize `reformat_one` to handle the `--code` case too. 

Thanks for being patient with me!

File: src/black/__init__.py
Comment (ichard26) [inline src/black/__init__.py:403]: Also technically misleading but I can't think of anything better.

**Comment by HassanAbouelela** [comment]
Thanks for the feedback, and I totally agree. Originally I wanted to maintain the whole “thrown together, mainly for testing” behavior, but as the PR has progressed, the feature has become quite bloated, and repeated a lot of the same code from the other functions. I’ll see what I can do to formalize this flag a bit more.

**Comment by HassanAbouelela** [comment]
Updated the code to use the STDIN formatter.

As a note for maintainers: we should probably squash/rebase this PR before merging. Most of the commits prior to switching to the STDIN formatter are irrelevant.

**Comment by JelleZijlstra** [comment]
We always squash merge, fyi. Also, could you fix the merge conflict in CHANGES.md?

**Comment by HassanAbouelela** [comment]
Thanks, good to know. The conflicts have been fixed.

File: src/black/__init__.py
Comment (JelleZijlstra) [inline src/black/__init__.py:539]: This will make it print something like "<all of your code> already well formatted". That's going to be pretty ugly if you input multiple lines. I'd suggest passing something like `Path("<string>")`, matching what Python itself does:
```
% python -c 'def f(x): 1/0
quote> f(1)
quote> '''
Traceback (most recent call last):
  File "<string>", line 2, in <module>
  File "<string>", line 1, in f
ZeroDivisionError: division by zero
```

File: src/black/report.py
Comment (JelleZijlstra) [inline src/black/report.py:7]: Prefer `from typing import Union` (but with my suggested change above you should be able to undo all changes to this file)

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks, this looks significantly cleaner

File: src/black/__init__.py
Comment (JelleZijlstra) [reply src/black/__init__.py]: Just applied this change myself so @cooperlees can do a release if he wants

**Comment by JelleZijlstra** [comment]
Just tried this out locally and decided that it's better to make `-c` apply `--quiet` by default because the extra "All done" output isn't really useful.

There might also be a discrepancy regarding whether `pyproject.toml` is used; looking into that more now.

**Comment by JelleZijlstra** [comment]
OK, there's indeed a problem with pyproject.toml:
```
(black) jelle@mbpt-root tmp % black -c "a = 'b'"  
a = "b"
(black) jelle@mbpt-root tmp % echo "a = 'b'" | black -
a = 'b'
All done! ✨ 🍰 ✨
1 file left unchanged.
(black) jelle@mbpt-root tmp % cat pyproject.toml 
[tool.black]
skip-string-normalization = true
```
I think the correct behavior is to pick up the pyproject.toml. That way, users can run `black -c` in their project directory to see how black would format code in that project.

The reason for the discrepancy is that in `black.read_pyproject_toml`, we pass `()` to `black.files.find_pyproject_toml`, and that function returns `Path("/").resolve()` if no src is provided. I think it should instead default to searching for a toml in the cwd. The current behavior would make it pick up `/pyproject.toml` at the file system root, which doesn't make much sense.

File: src/black/__init__.py
Comment (HassanAbouelela) [reply src/black/__init__.py]: Resolved in 39c3ae1.

File: src/black/report.py
Comment (HassanAbouelela) [reply src/black/report.py]: Resolved in 5f85ff5.

File: src/black/files.py
Comment (JelleZijlstra) [inline src/black/files.py:41]: I think this should be `srcs = [Path.cwd()]` to make this work like it does for `-`. In practice, this will make a difference if your pyproject.toml is in a parent directory.

Comment (HassanAbouelela) [reply src/black/files.py]: That makes sense, updated along with the tests.

Comment (ichard26) [reply src/black/files.py]: Unfortunately now tests are failing on Python 3.6 and 3.7. It's not clear what the issue is from an initial lookover. If you look into it that would be great but I wouldn't mind looking into it myself.

Comment (HassanAbouelela) [reply src/black/files.py]: I managed to replicate and found out it's due to the mock.call interface changing between [3.7](https://docs.python.org/3.7/library/unittest.mock.html#call) and [3.8](https://docs.python.org/3.8/library/unittest.mock.html#call). 3.8+ is backward compatible with the <3.8 format, so it was an easy fix.

This solution feels a bit cryptic, especially if you aren't familiar with the call structure, but it gets the job done. If anyone has better solutions or a different way to test this change, I'd be happy to implement it.

**Review by JelleZijlstra — APPROVED** [review]

**Review by ichard26 — APPROVED** [review]
I am happy with the final product for the most part. The only thing I'd dock against this patch is the general awkwardness of using `format_stdin_to_stdout` in `reformat_code`, but this is already quite complex so I'd rather refactor this (and other parts of the "main called to formatting will start" pathway) in a separate PR. 

Thank you so much for your contribution! This project is only possible by contributions like these :black_heart:. You're awesome, @HassanAbouelela! Congratulations on your first PR for psf/black. It's great to be able to get rid of that documentation warning for once and all. Also thanks for playing a role in the very nice welcome to Python Discord.

P.S. since you seem to be \*the\* person who manages to break all of our development workflows, it would be even more awesome if you could provide some feedback on your experience contributing here, we got a lot to learn (especially in the Windows department) so yeah. More details in #2238.

**Comment by cooperlees** [comment]
This needs the `CHNAGES.md` fixed ... Wonder how we can automate checking for that ...

**Comment by ichard26** [comment]
Oh fun! Probably automatically although that will make it harder to cleanly separate the logic into own repo when I *eventually* do that.

**Comment by cooperlees** [comment]
https://github.com/psf/black/commit/5de8c5f2f798d97133de3a1b391c04813c314c48

**Pull request merged**
